homeassistant/components/sensor/gtfs.py

```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
     @property
     def device_state_attributes(self):
         """Return the state attributes."""
-        return self._attributes
+        return {k: self._attributes[k] for k in sorted(self._attributes)}
 
     @property
     def icon(self):
         """Icon to use in the frontend, if any."""
+        self._icon = ICONS.get(self._route_type, ICON)
         return self._icon
 
     def update(self):
         """Get the latest data from GTFS and update the states."""
         with self.lock:
+            # Fetch valid stop information once
+            if not self._origin:
+                stops = self._pygtfs.stops_by_id(self.origin)
+                if not stops:
+                    raise ValueError(
+                        "Origin stop ID {} not found".format(self.origin))
+                self._origin = stops[0]
+
+            if not self._destination:
+                stops = self._pygtfs.stops_by_id(self.destination)
+                if not stops:
+                    raise ValueError("Destination stop ID {} not found".format(
+                        self.destination))
+                self._destination = stops[0]
+
+            # Fetch next departure
             self._departure = get_next_departure(
                 self._pygtfs, self.origin, self.destination, self._offset)
             if not self._departure:
+                self._agency = None
+                self._route = None
+                self._route_type = None
                 self._state = None
-                self._attributes = {'Info': 'No more departures today'}
+                self._trip = None
                 if self._name == '':
                     self._name = (self._custom_name or DEFAULT_NAME)
                 return
 
             self._state = self._departure['minutes_until_departure']
 
-            origin_station = self._departure['origin_station']
-            destination_station = self._departure['destination_station']
-            origin_stop_time = self._departure['origin_stop_time']
-            destination_stop_time = self._departure['destination_stop_time']
-            agency = self._departure['agency']
-            route = self._departure['route']
-            trip = self._departure['trip']
+            # Fetch metadata details once, unless updated
```

Comment:
Metadata objects all loaded from database here for later processing as attributes in `update_attributes()` below. Loading is done only once unless references have changed.